### PR TITLE
Adding initial scripts and data relating to the script for preserving consent to the dual-license

### PR DESCRIPTION
**Pull request by mchowning**
This PR is for persisting the grants of consent to the dual license from https://github.com/WordPress/gutenberg/issues/31893 and https://github.com/WordPress/gutenberg/issues/31913. 

* The `README.md` file includes an explanation for how the list of contributors was compiled as well as how to run the relevant scripts to preserve the latest relevant comments on the above issues.
* The `status.md` file is generated by the scripts that preserve the latest responses.
* The `scripts/` folder contains the relevant javascript files that can be executed (following the directions in the `README.md` file to preserve the latest responses.
* The `data/` folder includes the data files underlying the identification of all of Gutenberg's contributors as well as the `dual-license-responses.json` file which is where all of the relevant responses are being saved.

### CODE REVIEW

File: README.md
Comment (guarani) [inline README.md]: > more information available here

Is the "here" meant to be a link? Also, the next sentence starts "For more information" so it sounds like there's some redundancy there.

File: README.md
Comment (guarani) [inline README.md]: It looks like these links are not rendered properly. Since the `#` already makes a link out of these, would it make sense to remove the URLs?

File: README.md
Comment (guarani) [inline README.md]:
```suggestion
This script will automatically run [`summarize-dual-license-responses.js`](scripts/summarize-dual-license-responses.js), which prints out some current status of the requests for consent and updates the [`status.md`](status.md) file.
```

File: README.md
Comment (guarani) [inline README.md:71]:
```suggestion
Although this query captures logins of both the "author" and the person that merged each PR, the data shows that there is no one who has merged a PR who has not also authored at least one PR.
```
This change makes it clear that this refers to people who have merged a PR, not people who have merged PRs. Let me know if that makes sense.

Comment (guarani) [inline README.md]:
```suggestion
Although this query captures logins of both the "author" and the person that merged each PR, the data shows that there is no one who has merged a PR who has not also authored at least one PR. In other words, including logins from the "mergedBy" field does not add any additional contributors beyond those captured by the "author" field.
```
This clears an ambiguity I see where it wasn't clear if it meant "people who have merged one or more PRs" or "people who have one or more merged PRs".

File: README.md
Comment (guarani) [inline README.md]:
```suggestion
The list of all "authors" on all these commits included every "author" returned here, so we are fine just including everyone listed as in the "authors" object in the above response:
```
If I'm reading this right, it should say commits not PRs here, right?

File: README.md
Comment (guarani) [inline README.md]:
```suggestion
The json has two top level nodes. The `claimedEmails` node includes an object in `claimedEmails.responses` for each email address that has contributed to Gutenberg's `trunk` branch (i.e., by being an author or co-author of a commit on that branch) but was not associated with any GitHub account. 
```

File: scripts/update-dual-license-responses.js
Comment (guarani) [inline scripts/update-dual-license-responses.js]:
```suggestion
    console.log(`\nThis script doesn't persist any changes to the ${jsonFileName} file until the very end of the script.`)
```

File: scripts/utils.js
Comment (guarani) [inline scripts/utils.js]:
```suggestion
            " that have been reviewed and do not need to be reviewed again. The 'responses'" +
```

File: scripts/utils.js
Comment (guarani) [inline scripts/utils.js]:
```suggestion
            " that have been reviewed and do not need to be reviewed again. The 'preservedResponses'" +
```

File: README.md
Comment (guarani) [inline README.md]:
```suggestion
}
```

File: README.md
Comment (guarani) [inline README.md]:
```suggestion
These users were extracted from the "co-authors" of any PRs on trunk (every commit "author" was covered by the following "coauthors" query), where the email did not have an associated github account (the "user" was null).
```

File: README.md
Comment (guarani) [inline README.md]: Is this file meant to be added to the repo?

**Comment by guarani** [comment]
Hey @mchowning, I've spent some time playing with the queries, running scripts, reading the documentation and looking at the code. This is looking great and I didn't see any major issues.

Here are some potential issues I thought of before reviewing:
1. **Do we handle users editing their comment (e.g. changing their mind with regard to their consent)?** Looking at the scripts, when the script is re-run, we ignore comments we've already handled. This is probably OK because we record the timestamp of the original consent and the original consent can probably be found in the comment's edit history.
2. Do we handle contributors editing others' comments? Same as above.
3. **Do we handle non-contributors users leaving "consenting comments"?**: I think this is possible but wouldn't matter much. We just need to keep it in mind especially when tallying numbers.
4. **Do we handle contributors commenting multiple times?**: Again this is possible I think, but shouldn't really matter.
5. **Do we have to verify somehow the commit ownership claims made in https://github.com/WordPress/gutenberg/issues/31913?**: I'm not sure about this one, if it is possible for people to (inadvertently or otherwise) claim someone else's commits, is that something we need to handle? 

I'll come back to review again once you've had a chance to reply to the above comments and suggestions.

File: README.md
Comment (mchowning) [reply README.md]: I decided to expand that into a full sentence (which follows) and forgot to take out the parenthetical. Thanks for catching that. I'll remove the parenthetical.

File: README.md
Comment (mchowning) [reply README.md]: Nope, that was an oversight. I'll update this to be the `all_commits_before_2021-04-16.json` file.

File: README.md
Comment (mchowning) [reply README.md]: Oh, they're not rendering because I reversed the `()` and `[]`. I fixed that and I think it's good now. I don't think I can just use the #31893 text by itself to create the link because those are issues in a different repository, so GitHub won't connect them up.

**Comment by mchowning** [comment]
Thanks for the close review of this @guarani , I really appreciate it! 🙇 

> Do we handle users editing their comment (e.g. changing their mind with regard to their consent)? Looking at the scripts, when the script is re-run, we ignore comments we've already handled.

The consent we're requesting (and preserving) here is irrevocable ("I irrevocably consent..."), so we don't need to handle users withdrawing their consent because the consent remains valid.

> This is probably OK because we record the timestamp of the original consent and the original consent can probably be found in the comment's edit history.

Just noting that this seems to not work for deleted comments (I did not find a good way to recover them), which is one of the reasons we wanted to preserve the grants of consent and not just rely on GitHub to maintain that info for us.

> Do we handle contributors editing others' comments? Same as above.

The only way I'm guarding against this is by frequently checking the comments and preserving them. I checked and there have been no edits to this point that changed the meaning of the comment, but it might be nice to flag comments with edits before preserving them in the future. I'll think about this.

> Do we handle non-contributors users leaving "consenting comments"?: I think this is possible but wouldn't matter much. We just need to keep it in mind especially when tallying numbers.

If a user who is not a contributor leaves a comment, then the script just [skips it](https://github.com/WordPress/gutenberg-license/blob/initial-scripts/scripts/update-dual-license-responses.js#L165-L167).

> Do we handle contributors commenting multiple times?: Again this is possible I think, but shouldn't really matter.

If this happens (and it happened far more than I anticipated it would), and the earlier comment granted consent, [then the script reports the new comment and asks if we want to preserve the new comment instead of the previous one](https://github.com/WordPress/gutenberg-license/blob/initial-scripts/scripts/update-dual-license-responses.js#L135-L148).

> Do we have to verify somehow the commit ownership claims made in WordPress/gutenberg#31913?: I'm not sure about this one, if it is possible for people to (inadvertently or otherwise) claim someone else's commits, is that something we need to handle?

We don't have a way to independently verify that (if we did, we wouldn't need to ask them if they owned the commits). I think we're fine though because I am pinging those people because there is _very_ strong evidence that they are the owners, so it's not like some random person is claiming ownership. If we end up getting ownership claims from other people who do not otherwise appear to be the likely owners, then we would want to be more cautious about just accepting their claims.

File: README.md
Comment (guarani) [reply README.md]: Sounds good! I can't mark this as resolved so I'll leave that up to you.

File: README.md
Comment (guarani) [reply README.md]: Makes sense 👍

File: README.md
Comment (guarani) [reply README.md]: No worries 👍

**Comment by guarani** [comment]
I just noticed there are some emails such as `jason@o.lan` and `jason@readysetandco.com` that don't appear in the table on https://github.com/WordPress/gutenberg/issues/31913 but are in `claimedEmails.responses`, is that expected? It looks like Jason replied [here](https://github.com/WordPress/gutenberg/issues/31913#issuecomment-843408185), so while it looks like he was pinged, I'm not quite connecting the dots here.

**Comment by mchowning** [comment]
> I just noticed there are some emails such as jason@o.lan and jason@readysetandco.com that don't appear in the table on WordPress/gutenberg#31913 but are in claimedEmails.responses, is that expected? It looks like Jason replied here, so while it looks like he was pinged, I'm not quite connecting the dots here.

Good catch! There were some concerns expressed about "oversharing" these emails because the owners of those emails might have wanted them to be private. Although each of those emails came from a public commit (so they actually already are public), I'm manually removing the relevant rows from the table once there is a response relating to that email to reduce the "oversharing." That's why you're seeing that the emails/users with responses are no longer in the table.

**Comment by mchowning** [comment]
> Do we handle contributors editing others' comments? 

:wave: @guarani ! I added a check in the script that pauses and gives a warning if a comment has an editor other than the comments author in https://github.com/WordPress/gutenberg-license/pull/1/commits/2e784d72cf703810b25ade42f585df8d6949023d. We'll just need to manually check those comments to make sure there is nothing weird going on (fortunately, this doesn't happen very often, particularly if we run this script frequently, since we catch most comments before they have much time to get edited). I think that covers this scenario, but wdyt?

If the author is the person that edited the comment, then we're fine preserving the most recent edit. Only if they are denying consent might we want to go back and check the edits to make sure none of them included an irrevocable grant of consent. 

And now that I think about it, we should probably go back later and double-check that no one denying consent went back and edited their comment to grant consent. That would be a nice surprise.

**Comment by guarani** [comment]
> > Do we handle contributors editing others' comments?
> 
> 👋 @guarani ! I added a check in the script that pauses and gives a warning if a comment has an editor other than the comments author in [2e784d7](https://github.com/WordPress/gutenberg-license/commit/2e784d72cf703810b25ade42f585df8d6949023d). We'll just need to manually check those comments to make sure there is nothing weird going on (fortunately, this doesn't happen very often, particularly if we run this script frequently, since we catch most comments before they have much time to get edited). I think that covers this scenario, but wdyt?

Sounds good! I think that covers it.

> If the author is the person that edited the comment, then we're fine preserving the most recent edit. Only if they are denying consent might we want to go back and check the edits to make sure none of them included an irrevocable grant of consent.

👍 

> And now that I think about it, we should probably go back later and double-check that no one denying consent went back and edited their comment to grant consent. That would be a nice surprise.

True, there were only two but 🤞 

---

I think this is ready to merge 🙇 Thanks @mchowning!

**Comment by mchowning** [comment]
> I think this is ready to merge

🎉  Thanks @guarani ! Can you "approve" the PR? I just added branch protections for `trunk`, so I can't merge without an approval. I _think_ I'll also need to add you to this repository for your approval to satisfy the branch protection rule, but I wanted to go ahead and get the approval first to see if maybe that isn't necessary.

**Review by guarani — APPROVED** [review]
Looks good to me, @mchowning! Great work ❤️

**Comment by guarani** [comment]
> At least 1 approving review is required by reviewers with write access. 

Looks like you'll need to add me to the repo too as you mention above.

**Pull request merged**
